pkg/server/conn_stmt.go

```diff
@@ -331,97 +336,143 @@ func (cc *clientConn) executePreparedStmtAndWriteResult(ctx context.Context, stm
+	failpoint.Inject("avoidEagerCursorFetch", func() {
+		failpoint.Return(false, false, errors.New("failpoint avoids eager cursor fetch"))
+	})
+	crs := resultset.WrapWithCursor(rs)
+	cc.initResultEncoder(ctx)
+	defer cc.rsEncoder.Clean()
+	// fetch all results of the resultSet, and stored them locally, so that the future `FETCH` command can read
+	// the rows directly to avoid running executor and accessing shared params/variables in the session
+	// NOTE: chunk should not be allocated from the connection allocator, which will reset after executing this command
+	// but the rows are still needed in the following FETCH command.
+
+	// create the row container to manage spill
+	// this `rowContainer` will be released when the statement (or the connection) is closed.
+	rowContainer := chunk.NewRowContainer(crs.FieldTypes(), vars.MaxChunkSize)
+	rowContainer.GetMemTracker().AttachTo(vars.MemTracker)
+	rowContainer.GetMemTracker().SetLabel(memory.LabelForCursorFetch)
+	rowContainer.GetDiskTracker().AttachTo(vars.DiskTracker)
+	rowContainer.GetDiskTracker().SetLabel(memory.LabelForCursorFetch)
+	if variable.EnableTmpStorageOnOOM.Load() {
+		failpoint.Inject("testCursorFetchSpill", func(val failpoint.Value) {
+			if val, ok := val.(bool); val && ok {
+				actionSpill := rowContainer.ActionSpillForTest()
+				defer actionSpill.WaitForTest()
+			}
+		})
+		action := memory.NewActionWithPriority(rowContainer.ActionSpill(), memory.DefCursorFetchSpillPriority)
+		vars.MemTracker.FallbackOldAndSetNewAction(action)
+	}
+	defer func() {
+		if err != nil {
+			rowContainer.GetMemTracker().Detach()
+			rowContainer.GetDiskTracker().Detach()
+			errCloseRowContainer := rowContainer.Close()
+			if errCloseRowContainer != nil {
+				logutil.Logger(ctx).Error("Fail to close rowContainer in error handler. May cause resource leak",
+					zap.NamedError("original-error", err), zap.NamedError("close-error", errCloseRowContainer))
 			}
 		}
+	}()
+
+	for {
+		chk := crs.NewChunk(nil)
 
-		reader := chunk.NewRowContainerReader(rowContainer)
-		crs.StoreRowContainerReader(reader)
-		stmt.StoreResultSet(crs)
-		stmt.StoreRowContainer(rowContainer)
-		if cl, ok := crs.(resultset.FetchNotifier); ok {
-			cl.OnFetchReturned()
+		if err = crs.Next(ctx, chk); err != nil {
+			return false, false, err
+		}
+		rowCount := chk.NumRows()
+		if rowCount == 0 {
+			break
 		}
-		stmt.SetCursorActive(true)
-		defer func() {
-			if err != nil {
-				reader.Close()
 
-				// the resultSet and rowContainer have been closed in former "defer" statement.
-				stmt.StoreResultSet(nil)
-				stmt.StoreRowContainer(nil)
-				stmt.SetCursorActive(false)
-			}
-		}()
+		err = rowContainer.Add(chk)
+		if err != nil {
+			return false, false, err
+		}
+	}
 
-		if err = cc.writeColumnInfo(crs.Columns()); err != nil {
-			return false, err
+	reader := chunk.NewRowContainerReader(rowContainer)
+	defer func() {
+		if err != nil {
+			reader.Close()
 		}
+	}()
+	crs.StoreRowContainerReader(reader)
```

Comment:
Seems `StoreRowContainerReader` is the only extra method for `EagerCursorResultSet` and is only used here. Why not to remove `EagerCursorResultSet` and add a second argument `reader` to `WrapWithCursor`.

We can rename `WrapWithCursor` to `WrapWithRowContainerCursor`

Reply:
Done.